src/frontends/lean/elaborator.cpp

```diff
@@ -2884,350 +2869,405 @@ expr elaborator::instantiate_mvars(expr const & e, std::function<bool(expr const
-            expected_type = none_expr();
-    }
-    bool use_subobjects = !m_opts.get_bool("old_structure_cmd", false);
-    auto info = get_structure_instance_info(e);
-    name & S_name = info.m_struct_name;
-    buffer<name> & fnames = info.m_field_names;
-    buffer<expr> & fvalues = info.m_field_values;
-    bool catchall = info.m_catchall;
-
-    if (!S_name.is_anonymous() && !is_structure(env(), S_name))
-        throw elaborator_exception(e, sstream() << "invalid structure instance, '" <<
-                                   S_name << "' is not the name of a structure type");
-    lean_assert(fnames.size() == fvalues.size());
-    struct source { expr m_e; name m_S_name; };
-    buffer<source> sources;
-    for (expr src : info.m_sources) {
-        lean_assert(!m_in_pattern);
-        src = visit(src, none_expr());
-        synthesize_type_class_instances();
-        expr type  = instantiate_mvars(whnf(infer_type(src)));
-        expr src_S = get_app_fn(type);
-        if (!is_constant(src_S) || !is_structure(m_env, const_name(src_S))) {
-            auto pp_fn = mk_pp_ctx();
-            report_or_throw(elaborator_exception(e,
-                                       format("invalid structure notation source, not a structure") +
-                                       pp_indent(pp_fn, src) +
-                                       line() + format("which has type") +
-                                       pp_indent(pp_fn, type)));
-        } else {
-            sources.push_back(source {copy_tag(src, mk_as_is(src)), const_name(src_S)});
-        }
+/** Elaborate the structure instance notation `{... := ..., ...}` */
+class visit_structure_instance_fn {
+    // an elaborated source `..(m_e : m_S_name ...)`
+    struct source {
+        expr m_e;
+        // known to be a structure
+        name m_S_name;
+    };
+
+    elaborator & m_elab;
+    // note: fields needed by trace macros
+    environment & m_env = m_elab.m_env;
+    type_context & m_ctx = m_elab.m_ctx;
+
+    // inputs
+    expr m_e, m_ref = m_e;
+    optional<expr> m_expected_type;
+    bool m_use_subobjects = !m_elab.m_opts.get_bool("old_structure_cmd", false);
+    structure_instance_info m_info = get_structure_instance_info(m_e);
+    name & m_S_name = m_info.m_struct_name;
+    buffer<name> const & m_fnames = m_info.m_field_names;
+    buffer<expr> const & m_fvalues = m_info.m_field_values;
+
+    // set of fnames that have been used by `create_field_mvars`. Not using all of them is an error.
+    name_set m_fnames_used;
+    // elaborated sources
+    buffer<source> m_sources;
+
+    /* Because field default values may introduce arbitrary dependencies on other field values (not only within a structure,
+     * but even between a structure and its parent structures, in either direction), we cannot elaborate fields in a static
+     * order. Instead, we register a metavar and an elaboration function for each field (create_field_mvars).
```

Comment:
@leodemoura These registered elaboration functions were a major refactoring, but I'm very happy with the result. It reduces some horrible duplication and makes the overall approach much more general and less brittle. E.g. you could now have an inst implicit field whose resolution depends on the value of another field, if you really wanted to.